crates/turborepo-lib/src/shim.rs

```diff
@@ -0,0 +1,260 @@
+    /// kick over to the local turbo installation.
+    ///
+    /// # Arguments
+    ///
+    /// * `turbo_state`: state for current execution
+    ///
+    /// returns: Result<i32, Error>
+    fn run_correct_turbo(self, shim_args: ShimArgs) -> Result<Payload> {
+        let local_turbo_path = self.root.join("node_modules").join(".bin").join({
+            #[cfg(windows)]
+            {
+                "turbo.cmd"
+            }
+            #[cfg(not(windows))]
+            {
+                "turbo"
+            }
+        });
+
+        let current_turbo_is_local_turbo = local_turbo_path == current_exe()?;
+        // If the local turbo path doesn't exist or if we are local turbo, then we go
+        // ahead and run the Go code linked in the current binary.
+        if current_turbo_is_local_turbo || !local_turbo_path.try_exists()? {
+            cli::run()
+        } else {
+            // Otherwise we spawn the local turbo process.
+            Ok(Payload::Rust(
+                self.spawn_local_turbo(&local_turbo_path, shim_args),
+            ))
+        }
+    }
+
+    fn spawn_local_turbo(&self, local_turbo_path: &Path, mut shim_args: ShimArgs) -> Result<i32> {
+        let cwd = self
+            .root
+            .to_str()
+            .ok_or_else(|| anyhow!("Root directory path is invalid unicode"))?
+            .to_string();
+
+        let mut raw_args: Vec<_> = vec!["--skip-infer".to_string(), "--cwd".to_string(), cwd];
+        let has_single_package_flag = shim_args.single_package;
+
+        if self.mode == RepoMode::SinglePackage && !has_single_package_flag {
+            raw_args.push("--single-package".to_string());
+        }
+
+        raw_args.append(&mut shim_args.remaining_args);
+
+        // We spawn a process that executes the local turbo
+        // that we've found in node_modules/.bin/turbo.
+        let mut command = process::Command::new(local_turbo_path)
+            .args(&raw_args)
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .spawn()
+            .expect("Failed to execute turbo.");
+
+        Ok(command.wait()?.code().unwrap_or(2))
+    }
+}
+
+/// Checks for `TURBO_BINARY_PATH` variable. If it is set,
+/// we do not do any inference, we simply run the command as
+/// the current binary. This is due to legacy behavior of `TURBO_BINARY_PATH`
+/// that lets users dynamically set the path of the turbo binary. Because
+/// inference involves finding a local turbo installation and executing that
+/// binary, these two features are fundamentally incompatible.
+fn is_turbo_binary_path_set() -> bool {
+    env::var("TURBO_BINARY_PATH").is_ok()
+}
+
+pub fn run() -> Result<Payload> {
+    let args = ShimArgs::parse()?;
+
+    if args.skip_infer || args.has_help_flag || is_turbo_binary_path_set() {
+        return cli::run();
+    }
+
+    let repo_state = RepoState::infer(&args.cwd)?;
+
+    repo_state.run_correct_turbo(args)
```

Comment:
I think we should be executing current binary if we can't infer the repo. This will allow `--version` `--help` and `login` etc to work outside of a repo.
```suggestion
    if args.skip_infer || is_turbo_binary_path_set() {
        return cli::run();
    }

    if let Ok(repo_state) = RepoState::infer(&args.cwd) {
        repo_state.run_correct_turbo(args)
    }

    cli::run()
```